---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# variantscanR

<!-- badges: start -->
<img src="https://raw.githubusercontent.com/FrederiqueBoeykens/images/main/logo_sparkle.png" alt="Workflow" width="400">
<!-- badges: end -->

## Overview

Since the introduction of next-generation sequencing (NGS) techniques, whole-exome sequencing (WES) and whole-genome sequencing (WGS) not only revolutionized research, but also diagnostics. At present time, most of the tools available for the evaluation of NGS data are solely applicable for human analysis. On top of this, Veterinary Medicine also faces the additional problem that genetic diversity can be dangerously low, especially in the dog. For this reason, we developed _variantscanR_, an _R-package_, for the easy and straightforward identification of known disease-causing variants from a large collection of variants present in dogs and other animals.

The R-package variantscanR enables the filtering of variant call format (VCF file) files for the presence
of known disease-causing variants. In addition to the main variant-filtering function, the package allows
for the estimation of genetic diversity using both single and multisample VCF files. User-defined, file
format-specific options are available for the quality control of certain parameters using online database
screening. Finally, next to the filtering of known disease-causing variants, an additional step is
optional to gather the remaining variants located within the genes of interest. 


* `vcfscanneR()` Uploads a single sample VCF file (from a multi sample VCF file) into R.
* `chromosomenameR()` Changes 'NC_xxx' chromosome nomenclature to 'Chr' nomenclature.
* `annotateR_NCBI()` Changes the gene names of an NCBI BED file.
* `annotateR_ensembl()` Changes the gene names of an ensembl BED file.
* `pRocess()` Optional function that processes the user-defined file with the variants of interest. If desired, this function performes some quality control steps. This quality control might include providing a reference genome assembly (refseq argument, e.g. CanFam3.1) and removes all variants from other assemblies.
* `variantfiltR()` Filtering function that takes the single sample VCF file and filters for present variants of
#' interest.
* `extrafiltR()` Filtering function that takes the single sample VCF file and filters for present variants in genes of
#' interest. These genes of interest are obtained from the user-defined variants file.
* `diveRsity()` Provides a measure of diversity based on the percentage of heterozygosity per sample.


## Workflow
![Workflow](https://raw.githubusercontent.com/FrederiqueBoeykens/images/main/Figure1.png)

## Installation

```{r, eval = FALSE}
# Ensure that the 'devtools' package is installed in your R environment. If you haven't installed it yet, run the following command:
install.packages("devtools")

# Load the 'devtools' package by running the following command:
library(devtools)

# Install the package from GitHub using the 'install_github' function. Provide the Github repository owner and the repository name as arguments. Run the following command:
install_github("FrederiqueBoeykens/variantscanR")

# The package will be downloaded from GitHub and installed in your R environment along with its dependencies. After the installation is complete, you can load and use the package in your R scripts or sessions. 
```
---
title: "_variantscanR_ vignette"
author: "Fréderique Boeykens"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{variantscanR vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
## Generic information

Package: variantscanR

Type: Package

Title: Clinical Bioinformatics Tool for Variant Filtering of Known Disease-Causing Mutations in Domestic Animals

Version: 0.0.1
Authors@R: c(person("Fréderique", "Boeykens", email = c("frederique.boeykens@ugent.be", "frederique.boeykens@gmail.com"), role = c("cre", "aut")),person("Bart", "Broeckx", email = "bart.broeckx@ugent.be", role = "aut"))
Maintainer: Fréderique Boeykens <frederique.boeykens@ugent.be>

Description: The R-package variantscanR enables the filtering of variant call format (VCF file) files for the presence of known disease-causing variants. 
    In addition to the main variant-filtering function, the package allows for the estimation of genetic diversity using both single and multisample VCF files. 
    User-defined, file-specific options are available for the quality control of certain parameters using online database screening. 
    Finally, next to the filtering of known disease-causing variants, an additional step is optional to gather the remaining variants located within the genes of interest. 
    The R package is not species-specific and parallel computing is possible for some functions, demanding more computational power. 
    
License: GPL-3

URL: https://github.com/FrederiqueBoeykens/variantscanR``

Encoding: UTF-8

LazyData: True

LazyDataCompression: bzip2

RoxygenNote: 7.1.2

Imports: 
    vcfR,
    dplyr,
    stats,
    rebus,
    stringr,
    Biostrings,
    ggplot2,
    ggrepel,
    R3port,
    svMisc
    
Suggests: 
    rmarkdown,
    knitr,
    BSgenome,
    BSgenome.Cfamiliaris.UCSC.canFam3,
    tibble
    
VignetteBuilder: knitr

Depends: 
    R (>= 2.10)
    

